### DISCUSSION TITLE
How to use in-memory database with existing database for testing

### DISCUSSION THREAD

**Discussion by fiatcode-gh** [simolus3/drift#2963] [Q&A]
I'm writing an integration test and I need to load my existing database (from assets) into in-memory database. How can I achieve this? For now I use a regular file then delete it in tearDown for workaround. Thanks!

```dart
// ...imports

LazyDatabase _openTestConnection(File dbFile) {
  return LazyDatabase(() async {
    // extract the pre-populated database file from assets
    final blob = await rootBundle.load('assets/db.sqlite3');
    final buffer = blob.buffer;
    await dbFile.writeAsBytes(
        buffer.asUint8List(blob.offsetInBytes, blob.lengthInBytes));

    return NativeDatabase.createInBackground(dbFile);
  });
}

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  final dbFile = File('test.db');
  late MyDatabase myDatabase;

  setUp(() {
    myDatabase = MyDatabase(_openTestConnection(dbFile));
  });

  // ... tests

  tearDown(() {
    myDatabase.close();
    dbFile.deleteSync();
  });
}

```

**Comment by simolus3** [comment]
I'm not aware of a way to construct an in-memory database from the underlying bytes. So unfortunately it looks like creating a temporary file, copying the asset into that and then opening that as a database is the only way to go about this.

If the constant writing and deleting for each test is a problem, you can perhaps solve the issue by only writing the database to disk once (and not deleting it after every test). Because once the database is on disk, you can create an in-memory copy of it:

```dart
return LazyDatabase(() async {
  final reference = sqlite3.open(dbFile.path);
  final copy = sqlite3.openInMemory();
  await reference.backup(copy, -1).drain();
  return NativeDatabase.opened(copy);
});
```
